### PR TITLE
Add settings for HTTP header limitations

### LINKED ISSUE #25216: Limitations on amount of query parameters
**Describe the unexpected behaviour**
I try to use parameterized queries for our software to prevent SQL injections over user variable fields.
The implementation worked so far.

As our software can create quite large or complicated queries,
we ran into a first limitation.

We have a query with ~800 parameters and get back an error when using `curl` (for debugging).

**How to reproduce**
* Which ClickHouse server version to use
  * 21.5.6.6
* Which interface to use, if matters
  * HTTP
* Non-default settings, if any
* Query
  * As the parameters were not cleaned from confidential data, I can sadly not provide example tables and queries.
  * If needed for further investigation, I would try to create a query, which uses enough parameters to trigger the error.

**Expected behavior**
There should not be a limit to the amount of query fields.

**Error message and/or stacktrace**

```
$ curl -v `cat url.txt` --data-binary @query.sql
> Host: clickhouse:8123
> User-Agent: curl/7.64.0
> Accept: */*
> Content-Length: 49780
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
* Expire in 1000 ms for 0 (transfer 0x558051040fb0)
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 500 Internal Server Error
< Date: Fri, 11 Jun 2021 14:02:47 GMT
< Connection: Keep-Alive
< Content-Type: text/plain; charset=UTF-8
< X-ClickHouse-Server-Display-Name: 60601d5e39b6
< Transfer-Encoding: chunked
< X-ClickHouse-Exception-Code: 1000
<
Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = HTML Form Exception: Too many form fields (version 20.10.3.30 (official build))
* Connection #0 to host clickhouse left intact
```

How large is the limit on the query parameters and would there be a technical solution to not have an upper limit on the parameter count?

Thank you in advance.

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add settings `http_max_fields`, `http_max_field_name_size`, `http_max_field_value_size`

Detailed description / Documentation draft:
These settings were hard-coded, now they are configurable in the server settings. Fixes #25216, #25139

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Ok, but let's rename `http_max_fields_number` to `http_max_fields` for consistency with the other settings.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -234,6 +234,9 @@ class IColumn;
     M(Seconds, http_send_timeout, DEFAULT_HTTP_READ_BUFFER_TIMEOUT, "HTTP send timeout", 0) \
     M(Seconds, http_receive_timeout, DEFAULT_HTTP_READ_BUFFER_TIMEOUT, "HTTP receive timeout", 0) \
     M(UInt64, http_max_uri_size, 1048576, "Maximum URI length of HTTP request", 0) \
+    M(UInt64, http_max_fields_number, 1000000, "Maximum number of fields in HTTP header", 0) \
+    M(UInt64, http_max_field_name_size, 1024, "Maximum length of field name in HTTP header", 0) \
+    M(UInt64, http_max_field_value_size, 8192, "Maximum length of field value in HTTP header", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:239]: .

File: src/Server/HTTP/HTMLForm.cpp [hunk 2]
```diff
@@ -35,38 +36,41 @@ const std::string HTMLForm::ENCODING_MULTIPART = "multipart/form-data";
 const int HTMLForm::UNKNOWN_CONTENT_LENGTH = -1;
 
 
-HTMLForm::HTMLForm() : field_limit(DFL_FIELD_LIMIT), value_length_limit(DFL_MAX_VALUE_LENGTH), encoding(ENCODING_URL)
+HTMLForm::HTMLForm(ContextPtr context)
```
Comment (alexey-milovidov) [inline src/Server/HTTP/HTMLForm.cpp:39]: Why don't pass all three parameters directly instead of Context?
Or pass Settings instead of Context.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by Felixoid** [comment]
Since some curl's versions, the limit is 1M for all the headers, so they are broken. I decrease the defaults to 128K in https://github.com/ClickHouse/ClickHouse/pull/51163

**Pull request merged**